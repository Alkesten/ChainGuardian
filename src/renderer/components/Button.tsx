import * as React from 'react';

interface Props {
    type: string;
    children: any;
}

export const Button: React.FunctionComponent<Props> = ({ type, children, ...props }) => (
    <button className={`button btn-${type}`} {...props}>
        {children}
    </button>
);
