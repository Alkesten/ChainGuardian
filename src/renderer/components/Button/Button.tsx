import * as React from "react";

export interface IButtonProps {
    // type: string;
    // children: any;
    disabled?: boolean;
    onClick?: () => {};
}
export interface IBaseButtonProps extends IButtonProps{
    buttonType: string;
}
const TYPES = {
    PRIMITIVE: "primitive",
    PRIMARY: "primary",
    SECONDARY: "secondary",
    INVERTED: "inverted",
    DESTRUCTIVE: "destructive",
};

const BaseButton: React.FunctionComponent<IBaseButtonProps> = ({
    children,
    disabled,
    onClick,
    buttonType,
}) => (<button className={"button " + buttonType} disabled={disabled} onClick={onClick}>{children}</button>);

export const ButtonPrimitive: React.FunctionComponent<React.PropsWithChildren<IButtonProps>> = ({children, disabled, onClick}) => {
    return(<BaseButton disabled={disabled} onClick={onClick} buttonType={TYPES.PRIMITIVE}>{children}</BaseButton>);
    
};

