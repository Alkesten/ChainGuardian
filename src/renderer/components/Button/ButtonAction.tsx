import React from "react";
import arrow from "../../assets/img/buttons/Backtab.svg";
import add from "../../assets/img/buttons/Add.svg";
import copyDefault from "../../assets/img/buttons/CopyDefault.svg";
import ReactTooltip from "react-tooltip";

export interface IActionButtonProps {
    onClick?: () => void;
}

export interface ICopyButtonProps extends IActionButtonProps {
    clicked?: boolean;
}

export const BackTab: React.FunctionComponent<IActionButtonProps> = ({onClick}) => {
    return(<button 
        className={"back-tab"} 
        onClick={onClick}>
        <img className={"icon"} src={arrow} />
    </button>);
};

export const CopyButton: React.FunctionComponent<ICopyButtonProps> = (props: ICopyButtonProps) => {
    return(<button
        className={"copy-button"} 
        onClick={props.onClick}>
        <ReactTooltip effect="solid" place="right"
            getContent={(): string =>{return props.clicked ? "Copied!" : "Copy All";}}
        />
        <img data-tip="Copy all" data-place="top" className={"icon copy"} src={copyDefault} />
    </button>);
};

export const AddButton: React.FunctionComponent<IActionButtonProps> = ({onClick}) => {
    return(<button 
        className={"add-button"} 
        onClick={onClick}>
        <img className={"icon"} src={add} />
    </button>);
};

export const BackButton: React.FunctionComponent<IActionButtonProps> = ({onClick}) => {
    return(<button 
        className={"back-button"} 
        onClick={onClick}>
        <img className={"icon"} src={arrow} />
    </button>);
};

