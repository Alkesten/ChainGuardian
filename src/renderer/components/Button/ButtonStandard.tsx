import * as React from "react";

export interface IButtonProps {
    focused?: boolean;
    disabled?: boolean;
    onClick?: () => void;
    large?: boolean;
}
export interface IBaseButtonProps extends IButtonProps{
    buttonType?: string;
}
enum TYPES {
    PRIMITIVE = "primitive",
    PRIMARY = "primary",
    SECONDARY = "secondary",
    INVERTED = "inverted",
    DESTRUCTIVE = "destructive",
}

const BaseButton: React.FunctionComponent<IBaseButtonProps> = ({
    children,
    disabled,
    focused,
    onClick,
    buttonType,
    large,
}) => (focused ? <button 
    className={`button btn-${buttonType} focused`} 
    disabled={disabled} 
    onClick={onClick}>
    <div className={`${large ? "div-large" : ""}`}>{children}</div>
</button> : <button 
    className={`button btn-${buttonType}`} 
    disabled={disabled} 
    onClick={onClick}>
    <div className={`${large ? "div-large" : ""}`}>{children}</div>
</button>);

function getButton(props: React.PropsWithChildren<IBaseButtonProps>, type: TYPES ): React.ReactElement {
    return(
        <BaseButton
            disabled={props.disabled}
            focused={props.focused}
            onClick={props.onClick}
            large={props.large}
            buttonType={type}
        >{props.children}
        </BaseButton>
    );
}

export const ButtonPrimitive: 
React.FunctionComponent<React.PropsWithChildren<IButtonProps>> = (props) => {
    return(
        getButton(props, TYPES.PRIMITIVE)
    );
};

export const ButtonPrimary:
React.FunctionComponent<React.PropsWithChildren<IButtonProps>> = (props) => {
    return(
        getButton(props, TYPES.PRIMARY)
    );
};

export const ButtonSecondary:
React.FunctionComponent<React.PropsWithChildren<IButtonProps>> = (props) => {
    return(
        getButton(props, TYPES.SECONDARY)
    );
};

export const ButtonInverted:
React.FunctionComponent<React.PropsWithChildren<IButtonProps>> = (props) => {
    return(
        getButton(props, TYPES.INVERTED)
    );
};

export const ButtonDestructive:
React.FunctionComponent<React.PropsWithChildren<IButtonProps>> = (props) => {
    return(
        getButton(props, TYPES.DESTRUCTIVE)
    );
};