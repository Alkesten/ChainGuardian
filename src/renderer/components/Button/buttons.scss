.button {
    width: 168px;
    height: 54px;
    color: $neutral-800;
    font-size: 1.1rem;
    box-shadow: 0px 8px 16px rgba(0, 0, 0, 0.1);
    border-radius: 45px;
    outline: transparent;

    &.btn-primitive {
        background: $neutral-200;

        &:hover {
            background: $neutral-100;
            box-shadow: 0px 16px 24px rgba(0, 0, 0, 0.16);
        }

        &:focus {
            background: $neutral-100;
            box-shadow: 0 0 0 4px #ffffff, 0 0 0 8px $neutral-600;
        }

        &:active {
            background: $neutral-400;
        }

        &:disabled {
            background: $neutral-200;
            opacity: 0.5;
        }
    }

    &.btn-primary {
        background: $secondary-green-400;
        color: $secondary-green-700;

        &:hover {
            background: $secondary-green-300;
            box-shadow: 0px 16px 24px rgba(0, 0, 0, 0.16);
        }

        &:focus {
            background: $secondary-green-300;
            box-shadow: 0 0 0 4px #ffffff, 0 0 0 8px $secondary-green-600;
        }

        &:active {
            background: $secondary-green-500;
        }

        &:disabled {
            background: $secondary-green-400;
            opacity: 0.5;
        }
    }

    &.btn-secondary {
        background: $primary-blue-400;
        color: $primary-blue-700;

        &:hover {
            background: $primary-blue-300;
            box-shadow: 0px 16px 24px rgba(0, 0, 0, 0.16);
        }

        &:focus {
            background: $primary-blue-300;
            box-shadow: 0 0 0 4px #ffffff, 0 0 0 8px $primary-blue-600;
        }

        &:active {
            background: $primary-blue-500;
        }

        &:disabled {
            background: $primary-blue-400;
            opacity: 0.5;
        }
    }

    &.btn-inverted {
        background: $system-success-100;
        color: $system-success-700;

        &:hover {
            background: $system-success-200;
            box-shadow: 0px 16px 24px rgba(0, 0, 0, 0.16);
        }

        &:focus {
            background: $system-success-200;
            box-shadow: 0 0 0 4px #ffffff, 0 0 0 8px $system-success-600;
        }

        &:active {
            background: $system-success-400;
        }
    }

    &.btn-destructive {
        background: $system-error-600;
        color: $system-error-100;

        &:hover {
            background: $system-error-500;
            box-shadow: 0px 16px 24px rgba(0, 0, 0, 0.16);
        }

        &:focus {
            background: $system-error-500;
            box-shadow: 0 0 0 4px #ffffff, 0 0 0 8px $system-error-700;
        }

        &:active {
            background: $system-error-700;
        }
    }
}
