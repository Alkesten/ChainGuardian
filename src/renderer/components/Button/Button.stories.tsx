import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { ButtonPrimitive, ButtonPrimary, ButtonSecondary, ButtonInverted, ButtonDestructive } from './ButtonStandard';
import { BackTab, BackButton, AddButton, CopyButton } from './ButtonAction';
import { withKnobs, text, boolean, number } from '@storybook/addon-knobs';

storiesOf('Button', module).add('Submit Primitive', () => {
    const content = text('button content', 'Submit');
    const value = boolean('focus', false);
    const largeButton = boolean('large button', false);
    return  <ButtonPrimitive large={largeButton} focused={value} >{content}</ButtonPrimitive>;
}).addDecorator(withKnobs);

storiesOf('Button', module).add('Submit Primary', () => {
    const content = text('button content', 'Submit');
    const value = boolean('focus', false);
    const largeButton = boolean('large button', false);
    return <ButtonPrimary large={largeButton} focused={value} >{content}</ButtonPrimary>;
}).addDecorator(withKnobs);

storiesOf('Button', module).add('Submit Secondary', () => {
    const content = text('button content', 'Submit');
    const value = boolean('focus', false);
    const largeButton = boolean('large button', false);
    return <ButtonSecondary large={largeButton} focused={value} >{content}</ButtonSecondary>;
}).addDecorator(withKnobs);

storiesOf('Button', module).add('Submit Inverted', () => {
    const content = text('button content', 'Submit');
    const value = boolean('focus', false);
    const largeButton = boolean('large button', false);
    return <ButtonInverted large={largeButton} focused={value} >{content}</ButtonInverted>;
}).addDecorator(withKnobs);

storiesOf('Button', module).add('Submit Destructive', () => {
    const content = text('button content', 'Submit');
    const value = boolean('focus', false);
    const largeButton = boolean('large button', false);
    return <ButtonDestructive large={largeButton} focused={value} >{content}</ButtonDestructive>;
}).addDecorator(withKnobs);

storiesOf('Button', module).add('Back Tab', () => {
    return <BackTab />;
});

storiesOf('Button', module).add('Back Button', () => {
    return <BackButton />;
});

storiesOf('Button', module).add('Add Button', () => {
    return <AddButton />;
});

storiesOf('Button', module).add('Copy Button', () => {
    return <CopyButton />;
});