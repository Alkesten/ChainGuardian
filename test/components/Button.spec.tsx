import * as React from "react";
import * as renderer from "react-test-renderer";

import {ButtonPrimitive} from "../../src/renderer/components/Button/ButtonStandard";

describe("ButtonPrimitive", () => {
    it("renders correctly", () => {
        const tree = renderer
            .create(<ButtonPrimitive>Submit</ButtonPrimitive>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
    it("renders disabled button", () => {
        const tree = renderer
            .create(<ButtonPrimitive disabled={true}>Submit</ButtonPrimitive>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
    it("renders focused button", () => {
        const tree = renderer
            .create(<ButtonPrimitive focused={true}>Submit</ButtonPrimitive>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});