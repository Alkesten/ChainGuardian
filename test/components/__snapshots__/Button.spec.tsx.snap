// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddButton renders correctly 1`] = `
<button
  className="add-button"
>
  <img
    className="icon"
    src="test-file-stub"
  />
</button>
`;

exports[`BackButton renders correctly 1`] = `
<button
  className="back-button"
>
  <img
    className="icon"
    src="test-file-stub"
  />
</button>
`;

exports[`BackTab renders correctly 1`] = `
<button
  className="back-tab"
>
  <img
    className="icon"
    src="test-file-stub"
  />
</button>
`;

exports[`ButtonDestructive renders correctly 1`] = `
<button
  className="button btn-destructive"
>
  <div
    className=""
  >
    Submit
  </div>
</button>
`;

exports[`ButtonDestructive renders disabled button 1`] = `
<button
  className="button btn-destructive"
  disabled={true}
>
  <div
    className=""
  >
    Submit
  </div>
</button>
`;

exports[`ButtonDestructive renders focused button 1`] = `
<button
  className="button btn-destructive focused"
>
  <div
    className=""
  >
    Submit
  </div>
</button>
`;

exports[`ButtonDestructive renders large button 1`] = `
<button
  className="button btn-primitive"
>
  <div
    className="div-large"
  >
    Submit
  </div>
</button>
`;

exports[`ButtonInverted renders correctly 1`] = `
<button
  className="button btn-inverted"
>
  <div
    className=""
  >
    Submit
  </div>
</button>
`;

exports[`ButtonInverted renders disabled button 1`] = `
<button
  className="button btn-inverted"
  disabled={true}
>
  <div
    className=""
  >
    Submit
  </div>
</button>
`;

exports[`ButtonInverted renders focused button 1`] = `
<button
  className="button btn-inverted focused"
>
  <div
    className=""
  >
    Submit
  </div>
</button>
`;

exports[`ButtonInverted renders large button 1`] = `
<button
  className="button btn-primitive"
>
  <div
    className="div-large"
  >
    Submit
  </div>
</button>
`;

exports[`ButtonPrimary renders correctly 1`] = `
<button
  className="button btn-primary"
>
  <div
    className=""
  >
    Submit
  </div>
</button>
`;

exports[`ButtonPrimary renders disabled button 1`] = `
<button
  className="button btn-primary"
  disabled={true}
>
  <div
    className=""
  >
    Submit
  </div>
</button>
`;

exports[`ButtonPrimary renders focused button 1`] = `
<button
  className="button btn-primary focused"
>
  <div
    className=""
  >
    Submit
  </div>
</button>
`;

exports[`ButtonPrimary renders large button 1`] = `
<button
  className="button btn-primitive"
>
  <div
    className="div-large"
  >
    Submit
  </div>
</button>
`;

exports[`ButtonPrimitive renders correctly 1`] = `
<button
  className="button btn-primitive"
>
  <div
    className=""
  >
    Submit
  </div>
</button>
`;

exports[`ButtonPrimitive renders disabled button 1`] = `
<button
  className="button btn-primitive"
  disabled={true}
>
  <div
    className=""
  >
    Submit
  </div>
</button>
`;

exports[`ButtonPrimitive renders focused button 1`] = `
<button
  className="button btn-primitive focused"
>
  <div
    className=""
  >
    Submit
  </div>
</button>
`;

exports[`ButtonPrimitive renders large button 1`] = `
<button
  className="button btn-primitive"
>
  <div
    className="div-large"
  >
    Submit
  </div>
</button>
`;

exports[`ButtonSecondary renders correctly 1`] = `
<button
  className="button btn-secondary"
>
  <div
    className=""
  >
    Submit
  </div>
</button>
`;

exports[`ButtonSecondary renders disabled button 1`] = `
<button
  className="button btn-secondary"
  disabled={true}
>
  <div
    className=""
  >
    Submit
  </div>
</button>
`;

exports[`ButtonSecondary renders focused button 1`] = `
<button
  className="button btn-secondary focused"
>
  <div
    className=""
  >
    Submit
  </div>
</button>
`;

exports[`ButtonSecondary renders large button 1`] = `
<button
  className="button btn-primitive"
>
  <div
    className="div-large"
  >
    Submit
  </div>
</button>
`;

exports[`CopyButton renders correctly 1`] = `
<button
  className="copy-button"
>
  <img
    className="icon copy"
    src="test-file-stub"
  />
  <img
    className="icon copyall"
    src="test-file-stub"
  />
  <img
    className="icon copied"
    src="test-file-stub"
  />
</button>
`;
